src/io/asyncsocketudp.c

```diff
@@ -601,3 +749,79 @@ MVMObject * MVM_io_socket_udp_async(MVMThreadContext *tc, MVMObject *queue,
 
     return (MVMObject *)task;
 }
+
+
+/* Creates a UDP socket as with MVM_io_socket_udp_async, but also
+ * subscribes to the specified multicast. */
+MVMObject * MVM_io_socket_udp_async_multicast(MVMThreadContext *tc, MVMObject *queue,
+                                    MVMObject *schedulee, MVMString *host,
+                                    MVMint64 port, MVMint64 flags, MVMint64 iface,
+                                    MVMString *ssm_host, MVMObject *async_type) {
+    MVMAsyncTask       *task;
+    MulticastSetupInfo *msi;
+    struct sockaddr    *host_addr = NULL;
+    struct sockaddr    *bind_addr = NULL;
+    struct sockaddr    *ssm_addr  = NULL;
+    char               *host_str  = NULL;
+    char               *ssm_str   = NULL;
+
+    /* Validate REPRs. */
+    if (REPR(queue)->ID != MVM_REPR_ID_ConcBlockingQueue)
+        MVM_exception_throw_adhoc(tc,
+            "asyncudp target queue must have ConcBlockingQueue REPR");
+    if (REPR(async_type)->ID != MVM_REPR_ID_MVMAsyncTask)
+        MVM_exception_throw_adhoc(tc,
+            "asyncudp result type must have REPR AsyncTask");
+    if ((flags & 0b11) != 2)
+        MVM_exception_throw_adhoc(tc,
+            "asyncudp flags must indicate multicast to call MVM_io_socket_udp_async_multicast");
+
+    /* Resolve hostnames. (Could be done asynchronously too.) */
+    if (host && IS_CONCRETE(host)) {
+        MVMROOT3(tc, queue, schedulee, async_type) {
+            /* Resolve for multicast for consistency, these will almost always be IP-num based */
+            host_addr = MVM_io_resolve_host_name(tc, host, port, MVM_SOCKET_FAMILY_UNSPEC, MVM_SOCKET_TYPE_DGRAM, MVM_SOCKET_PROTOCOL_ANY, 1);
+
+            /* Convert the addresses. Multicast routines in libuv require char[] */
+            /* XXX Handle case if host_addr.sin6_scope_id != 0 && host_addr.sin6_scope_id != iface */
+            host_str = (char*) MVM_calloc(1, sizeof(char) * UV_MAXHOSTNAMESIZE);
+            uv_ip_name((const struct sockaddr*) host_addr, host_str, UV_MAXHOSTNAMESIZE - 1);
```

Comment:
According to https://man7.org/linux/man-pages/man3/inet_ntop.3.html, which `uv_ip_name` uses, the `- 1` is not necessary here.